### PR TITLE
INSP: Refactor borrow checker

### PR DESCRIPTION
**Pull request by farodin91**

### CODE REVIEW

**Comment by vlad20012** [comment]
It is not just refactoring. It make the inspection less strict to solve #1667, isn't it? But I think we should just make type inference better... (don't sure it is the right vision, may be we should fix it as soon as possible and then deal with type inference)

**Comment by farodin91** [comment]
I try to fix it directly, but i want to support it for all reference mutable iterators

**Comment by vlad20012** [comment]
Real problem is not in type inference directly, it is about exact/fuzzy impl matching (look at `RsPreciseTraitMatchingTest`). If you want to fix it, I have some suggestions. (And as I see, @kumbayo is looking in the same direction with #1697))

**Comment by farodin91** [comment]
What do suggestions do you think of? I want in this PR only fix issue directly correlate to the borrow checker.

File: src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt [hunk 1]
```diff
@@ -125,6 +125,16 @@ class RsBorrowCheckerInspectionTest : RsInspectionsTestBase(RsBorrowCheckerInspe
         }
     """)
 
+    fun `test no highlight for mutable for loops`() = checkByText("""
+        fn test() {
+            let mut xs: Vec<Vec<usize>> = vec![vec![1, 2], vec![3, 4]];
+
+            for test in &mut xs {
+                <error>test</error>.push(0)
```
Comment (matklad) [inline src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt]: Hm, I think there shouldn't be an error here? At lest the following compiles fine:

```Rust
fn main() {
    let mut xs: Vec<Vec<usize>> = vec![vec![1, 2], vec![3, 4]];
    for test in &mut xs {
        test.push(0)
    }
}
```

Comment (matklad) [reply src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt]: If this is a test for the current behavior which should be fixed later, could you add a `TODO: this should not show error here` ?

Comment (farodin91) [reply src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt]: I add a note

**Comment by vlad20012** [comment]
Hmm, I think this test confused me. So it is really just a refactoring. So thanks!
bors r+

**Pull request merged**